demo/wasm/src/lib.rs

```diff
@@ -11,16 +20,68 @@ pub extern "C" fn _call(input_data: *mut u8, input_len: usize) -> i32 {
             //print(888);
             print(*input_data.offset(i as isize) as i32);
         }
+    }
+
+    return 0; //a * a;
+}
 
+use std::ffi::CStr;
+use std::os::raw::c_char;
+
+fn make_internal<'s, T: Deserialize<'s>>(data: *mut c_char) -> T {
+    let c_str = unsafe { CStr::from_ptr(data) };
+    let actual_str = c_str.to_str().unwrap(); // Don't unwrap ever in real life
+    serde_json::from_str(actual_str).unwrap() // OMG you're still doing it! Have you learned nothing?!
+}
+
+fn make_external<T: Serialize>(data: *mut c_char, params_len: usize, internal: T) -> i32 {
+    let json = serde_json::to_string(&internal).unwrap(); //same!
+                                                          //    let json = "fish".to_string();
+    let bytes = json.as_bytes();
+    let len = bytes.len();
+    for i in 0..len {
+        unsafe {
+            *data.offset(i as isize) = bytes[i] as i8;
+        }
     }
+    len as i32
+    //unimplemented!()
+}
 
-    return 0;//a * a;
+#[no_mangle]
+pub extern "C" fn test_dispatch(data: *mut c_char, params_len: usize) -> i32 {
+    let input = make_internal(data);
+    let output = test(input);
+    make_external(data, params_len, output)
```

Comment:
Hm, so you are basically just writing over the same memory that you got the input through... which in our case starts at 0. And you're not checking the length, which could be be longer for the output than the input.. I would not expect this to work in the general case.. I believe we will have to implement a malloc() function that at least just increments an index by the number of allocated bytes..

Reply:
yes added #65 to account for this.